### PR TITLE
MINIFICPP-1453: Verify Supported TLS Protocols

### PR DESCRIPTION
**Pull request by martinzink**
Currently org::apache::nifi:minifi::io::TLSSocket
(which, among other things, handles communications with C2)
only accept TLSv1.2 protocol (hardcoded behaviour)

Added integration tests to verify this behaviour
  TLSClientSocketSupportedProtocolsTest
    creates simple TLS enabled test servers with various protocols
    org::apache::nifi::minifi::io::TLSSocket only connects
    to the TLSv1.2 enabled server

  TLSServerSocketSupportedProtocolsTest
    creates an org::apache::nifi::minifi::io::TLSServerSocket
    various test clients try to connect to it,
    but only the TLSv1.2 enabled client succeeds

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (szaszm) [inline extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp:172]: [Use auto to avoid redundant repetition of type names](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es11-use-auto-to-avoid-redundant-repetition-of-type-names) (multiple occurences thoughout the PR, but I don't want to spam the same comment)

```suggestion
      const auto socket_context = std::make_shared<org::apache::nifi::minifi::io::TLSContext>(configuration_);
```

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (szaszm) [inline extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp:53]: Is this intentionally unix-only? On Windows, the socket type is `SOCKET`, and other parts of the networking code of the test are unix-only as well.

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (szaszm) [inline extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:52]: I would make `successfulConnection` `const` as well, since it's just a derived value from `status`.

To be honest, I would make as many things `const` as reasonably possible, but this is not common practice in most contributions.

**Comment by szaszm** [comment]
And thanks for your first contribution. The general workflow is that we merge after every comment is resolved, there are 2-3 approvals on the pull request, no conflicts (rebase as needed) and no related CI issues.

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (szaszm) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp:41]: When storing them, pass-by-value and move is better IMO, but pass-by-ref-to-const is fine, too.

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (szaszm) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp:144]: source: According to [CONTRIB.md](https://github.com/apache/nifi-minifi-cpp/blob/main/CONTRIB.md), contributions are expected to follow the Google (C++) Style guide, and it says "[Data members of classes, both static and non-static, are named like ordinary nonmember variables, but with a trailing underscore.](https://google.github.io/styleguide/cppguide.html#Variable_Names)".

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: No, I've forgot about windows, should be fixed now.

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: fixed by third commit in PR

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: Should be fixed by third commit in PR.
Copy paste mistakes... I've originally used SecureSocketGetTCPTest.cpp and IntegrationBase.h as template. Any chance that linter could also catch these kind of style-guide violations?

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: Agree, renamed and refactored it to be more readable.

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (lordgamez) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: Currently cpplint is run for the PRs which should catch a lot of violations to the style-guide, but unfortunately it cannot catch all violations like variable namings.

Comment (lordgamez) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: I'm not sure if other tools like A-Style or clang-format might be able to catch more violations, we need to check as possible options. Some static code analysis tools like CodeChecker allow custom checkers to be created that can be used even for variable name checks.

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: moved them inside asserts instead, because they are not expected to fail but any further testing is impossible without these succeeding

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (lordgamez) [inline extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp:92]: Minor indent issue

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (lordgamez) [inline extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:101]: Minor indent issues

**Review by lordgamez — APPROVED** [review]
LGTM, only a minor indent issue. Thanks for the contribution!

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: Fixed the indentations, thanks for the review

**Comment by szaszm** [comment]
The issue on the ubuntu-16.04-all CI run seems to be related, could you check that?

**Comment by martinzink** [comment]
There was a race condition which could cause one of the asserts to fail, fixed it.

**Review by szaszm — APPROVED** [review]

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (szaszm) [inline extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:80]: Is this a leftover? The result of this name resolution doesn't seem to be used later aside from the assertion.

**Review by szaszm — CHANGES_REQUESTED** [review]
Sorry for going back on my approve, but local testing revealed another issue:
```
      Test  #76: TLSServerSocketSupportedProtocolsTest .............Subprocess aborted***Exception:   0.12 sec
[2021-01-27 12:58:23.914] [org::apache::nifi::minifi::io::TLSSocket] [info] Listening on 0.0.0.0:38778 with backlog 3
TLSServerSocketSupportedProtocolsTest: /home/szaszm/nifi-minifi-cpp-2/extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:104: static SocketDescriptor SimpleSSLTestClient::openConnection(const char*, const char*): Assertion `sfd != INVALID_SOCKET' failed.
```
We should try to figure this out before merging.

**Comment by szaszm** [comment]
The issue is with the `socket` call. I fixed it and added extra logging in martinzink#5

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: yeah it was leftover, I removed it

**Review by szaszm — APPROVED** [review]

**Comment by szaszm** [comment]
I'm waiting for one more approval before merging due to the extent of changes.

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp:204]: This overrides how to handle the broken pipe signal (which will happen because we are testing that you cannot connect with outdated protocol). The default behavior for this signal is to end the process.
This workaround is already used by extensions/standard-processors/tests/integration/SecureSocketGetTCPTest.cpp (I used that class as template for these tests)

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: thanks, fixed it

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: fixed

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: fixed

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: agreed, changed it

File: extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSClientSocketSupportedProtocolsTest.cpp]: fixed it

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: changed it

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: Good catch, it came from extensions/standard-processors/tests/integration/SecureSocketGetTCPTest.cpp which I used as template, removed it nonetheless.

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: Thanks, looks way more readable. I included this with the other suggested changes in the latest commit.

**Review by fgerlits — APPROVED** [review]

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (adamdebreceni) [inline extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:286]: we could use a `unique_ptr` for the `server_socket_` (to make it clear that it is getting destroyed at the end of this object) and maybe even call an explicit `reset` on it in the destructor of `TLSServerSocketSupportedProtocolsTest ` (since currently the correctness hinges on the order of `server_socket_` and `is_running_`)

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (adamdebreceni) [inline extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:284]: why store these functions, it seems to me that they are passed by copy to `server_socket_`

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (adamdebreceni) [inline extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:274]: I would argue that the templates forwarding to this template are unnecessary and we could call this one directly from `verifyTLSServerSocketExclusiveCompatibilityWithTLSv1_2 `

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: Made it unique to be more clear, calling reset however seemed little excessive since unique_ptr destructor will be called and that achieves the same thing. I did however changed the destructor to set is_running_ to false. This will basically achieve that the external check function inside the socket will return false, after that the unique_ptr destructor will destroy the socket which will set its own inside check to false and it'll wait for the thread to finish.

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: fixed

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp:268]: fixed

File: extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp
Comment (adamdebreceni) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: well if in the future somebody switches up `is_running_` and `server_socket_`, `is_running_` will get destroyed first, and then the still running `server_socket_` could accidentally access the then destroyed `is_running_` (and setting `is_running_` in the destructor does not seem to mitigate this), but you are right that it is currently not an issue, maybe a comment then, that the order of these fields is important?

Comment (martinzink) [reply extensions/standard-processors/tests/integration/TLSServerSocketSupportedProtocolsTest.cpp]: I see your point, you are right, an explicit reset call in the destructor before the destruction of the members seems the best way. fixed

**Review by adamdebreceni — APPROVED** [review]

**Pull request closed without merging**
